runtime/tchannel_server.go

```diff
@@ -64,197 +84,364 @@ func (ncr netContextRouter) Handle(ctx netContext.Context, call *tchannel.Inboun
+		return errors.Wrapf(err, "Could not ensure arg2reader is empty for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
 	}
 	if err := treader.Close(); err != nil {
-		return errors.Wrapf(err, "could not close arg3reader is empty for inbound call: %s::%s", service, method)
+		LogError(c.endpoint.Logger, err, "Could not close arg2reader for inbound request")
+		return errors.Wrapf(err, "Could not close arg2reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
 	}
+	return nil
+}
 
-	structWireValue, err := resp.ToWire()
+// readReqBody reads request body from arg3
+func (c *tchannelInboundCall) readReqBody(ctx context.Context) (wireValue wire.Value, err error) {
+	// fail fast if timed out
+	if deadline, ok := ctx.Deadline(); ok && time.Now().After(deadline) {
+		err = context.DeadlineExceeded
+		return
+	}
+
+	treader, err := c.call.Arg3Reader()
+	if err != nil {
+		LogError(c.endpoint.Logger, err, "Could not create arg3reader for inbound request")
+		err = errors.Wrapf(err, "Could not create arg3reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	buf := GetBuffer()
+	defer PutBuffer(buf)
+	if _, err = buf.ReadFrom(treader); err != nil {
+		_ = treader.Close()
+		LogError(c.endpoint.Logger, err, "Could not read from arg3reader for inbound request")
+		err = errors.Wrapf(err, "Could not read from arg3reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	wireValue, err = protocol.Binary.Decode(bytes.NewReader(buf.Bytes()), wire.TStruct)
 	if err != nil {
-		// If we could not write the body then we should do something else
-		// instead.
+		c.endpoint.Logger.Warn("Could not decode arg3 for inbound request", zap.Error(err))
+		err = errors.Wrapf(err, "Could not decode arg3 for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	if err = EnsureEmpty(treader, "reading request body"); err != nil {
+		_ = treader.Close()
+		LogError(c.endpoint.Logger, err, "Could not ensure arg3reader is empty for inbound request")
+		err = errors.Wrapf(err, "Could not ensure arg3reader is empty for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	if err = treader.Close(); err != nil {
+		LogError(c.endpoint.Logger, err, "Could not close arg3reader for inbound request")
+		err = errors.Wrapf(err, "Could not close arg3reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
 
-		_ = call.Response().SendSystemError(errors.New("Server Error"))
+	// request received when arg3reader is closed
+	c.endpoint.metrics.Recvd.Inc(1)
```

Comment:
the request is recvd earlier. We return an error above this which should return an `err` via `handle()` and thus report an error in `finish()`

I think we want to emit recvd in `start()`

Reply:
Moving to back to where it was originally.